dashboard/app/models/SCHOOL_DATA_README.md

```diff
@@ -0,0 +1,189 @@
++------+--------------+-------+---------+----------+------+-------+---------+---------+
+6 rows in set (0.32 sec)
+```
+
+# SchoolInfo validation logic requires you to specify a district, state, type, and country even if you specify a School which already has that info.
+Ideally those fields would not be required if a school is provided. 
+It is likely that there are code paths and queries that depend on having the denormalized data in the SchoolInfo object directly. Overriding the accessor methods to get these fields directly from the School would allow us to remove the denormalized data without breaking code access. To avoid breaking any other queries we would need to understand what queries exist (against the prod DB as well as the reporting DB and Redshift) and update them.
+Inconsistency is now mostly fixed by checking in validation if there is a school id and, if so, copying the denormalized data from the School and  upgrading the validation type to full. We still need to eliminate places where we are sending in mismatched data in the first place. When that happens it gets logged to Honeybadger (https://app.honeybadger.io/projects/3240/faults/35428612). Once those errors are cleaned up, we ought to change school_info validation to fail if any other data is provided along with a school id.
+
+# Multiple rows for same school_id in school_infos
+When a school_info has an associated school there is no additional data in the school_info that isn’t in the school. You would expect there the be at most one school info for each school. This is not the case
+```
+select school_info_count, count(*) as number_of_schools from (select school_id, count(*) as school_info_count from school_infos where school_id is not null group by school_id) as counts group by school_info_count order by count(*) desc;
++-------------------+-------------------+
+| school_info_count | number_of_schools |
++-------------------+-------------------+
+|                 1 |             34923 |
+|                 2 |             12793 |
+|                 3 |                17 |
+|                 4 |                 1 |
+|                 5 |                 1 |
++-------------------+-------------------+
+5 rows in set (0.15 sec)
+```
+There seem to be two main explanations for why we have multiple rows for the same school:
+* The school_info deduplication logic uses the validation type as part of the match. This means that if you are trying to create a school_info for a school using full validation it will not match an existing row for that school with validation type ‘none’. This issue is effectively fixed going forward since we no longer create school_infos with validation type ‘none’ when there is an associated school.
+* Deduplication needs to be explicitly called (it isn’t part of the model.) When it isn’t called we can create a additional row. This is the most likely explanation for the cases where we have more than two rows for a given school.
```

Comment:
Is the source of this multiple/conflicting info because two teachers from the same school might tell us different things about their schools?  Or perhaps a form issue where a teacher manually filled in details in the past, but later went back and picked a school from the autocomplete dropdown?

Reply:
I never really figured out the root cause of all the duplicates. For most of the schools with exactly two rows it seems to be that deduplication checked the validation type so we'd end up with one row with validation none and one with validation full. I'm not sure about the cases where there are 3 or more school_infos for the same school. Some of those rows have all columns the same. Possibly that is caused by duplicate form submissions? Deduplication isn't transactional so if two submissions are processed concurrently then they can both end up creating a row.

I'll add this info to the readme.

Reply:
Added.